jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala

```diff
@@ -308,59 +335,148 @@ private class Watches private(
+      val accepted = r.nextDouble() <= trainTestRatio
+      if (!accepted) {
+        labeledPointGroup.foreach(labeledPoint => {
+          testPoints += labeledPoint
+          testBaseMargins += labeledPoint.baseMargin
+        })
+        testGroups += labeledPointGroup.length
+      } else {
+        labeledPointGroup.foreach(trainBaseMargins += _.baseMargin)
+        trainGroups += labeledPointGroup.length
+      }
       accepted
     }
 
-    val (trainIter1, trainIter2) = trainPoints.duplicate
-    val trainMatrix = new DMatrix(trainIter1, cacheDirName.map(_ + "/train").orNull)
-    val trainGroups = buildGroups(trainIter2.map(_.group).toSeq).toArray
-    trainMatrix.setGroup(trainGroups)
+    val trainPoints = trainLabelPointGroups.flatMap(_.iterator)
+    val trainMatrix = new DMatrix(trainPoints, cacheDirName.map(_ + "/train").orNull)
+    trainMatrix.setGroup(trainGroups.result())
 
-    val testMatrix = new DMatrix(testPoints.iterator, cacheDirName.map(_ + "/test").orNull)
+    val testMatrix = new DMatrix(testPoints.result().iterator, cacheDirName.map(_ + "/test").orNull)
     if (trainTestRatio < 1.0) {
-      val testGroups = buildGroups(testPoints.map(_.group)).toArray
-      testMatrix.setGroup(testGroups)
+      testMatrix.setGroup(testGroups.result())
     }
 
-    r.setSeed(seed)
-    for (baseMargins <- baseMarginsOpt) {
-      val (trainMargin, testMargin) = baseMargins.partition(_ => r.nextDouble() <= trainTestRatio)
-      trainMatrix.setBaseMargin(trainMargin)
-      testMatrix.setBaseMargin(testMargin)
-    }
+    val trainMargin = fromBaseMarginsToArray(trainBaseMargins.result().iterator)
+    val testMargin = fromBaseMarginsToArray(testBaseMargins.result().iterator)
+    if (trainMargin.isDefined) trainMatrix.setBaseMargin(trainMargin.get)
+    if (testMargin.isDefined) testMatrix.setBaseMargin(testMargin.get)
 
     new Watches(trainMatrix, testMatrix, cacheDirName)
   }
 }
+
+
+/**
+  * Within each RDD partition, group the <code>XGBLabeledPoint</code> by group id.</p>
+  * While loading data into RDD, the items within a group may be break into the adjacent partitions.
+  * Need to throw away the fist and the last groups within a partition,
+  * since they may not be a complete group.</p>
+  * Make the first and last group as empty list, so they can be filtered out later easily.
+  * @param base collection of <code>XGBLabeledPoint</code>
+  */
+private[spark] class LabeledPointGroupIterator(base: Iterator[XGBLabeledPoint])
+  extends AbstractIterator[Array[XGBLabeledPoint]] {
+  private var tailPoint: XGBLabeledPoint = null
+
+  override def hasNext: Boolean = {
+    return base.hasNext
+  }
+
+  override def next(): Array[XGBLabeledPoint] = {
+    val builder = mutable.ArrayBuilder.make[XGBLabeledPoint]
+    var firstGroup = true
+    if (tailPoint != null) {
+      builder += tailPoint
+      firstGroup = false
+    }
+    var lastGroup = true
+    while (lastGroup == true && base.hasNext) {
```

Comment:
using lastGroup in this while loop is confusing....

I suggest you create another boolean variable indicating `nextGroup` 

or you may want to at least leave some comments about how you reuse lastGroup var

Reply:
one thing I am unclear about is that if, for example,

(1) Group 1's data is distributed to partition A and B and locates at the last group of A and first group of B, will we lose all data about Group A?

(2) if Group A is completely in partition A, will we lose all data about Group A?

Reply:
My current code blindly drop the first and last groups within a partition no matter they are complete group or not.

The way to keep the group and its ordering, is to group the items within a group before doing any shuffling or splitting.

If group1 is crossing partition A and partition B, there are three ways to deal group border issue:
1. stitch group1 by moving the items from partition B to partition A
2. simply drop the first and last groups in any partition
3. The input data is already in groups (Array[LabeledPoint]

Solution 1 needs pretty expensive Spark steps to fix the group crossing borders.
I picked the solution 2 here, because with distributed training, we are talking about millions or billions of training data, and to drop a few groups should have no impact on the accuracy of the model.

actually, in my private build, we directly generate the training data in groups, and I changed the entry method as "def trainDistributed(trainingData: RDD[Array[XGBLabeledPoint]], ...)"

I would like to know your thoughts on which way to go.

Reply:
Just a minor nitpick: `lastGroup == true` can be replaced with `lastGroup`.

Reply:
add one more flag to make code more readable: isNewGroup